src/compiler/program.ts

```diff
@@ -3508,11 +3526,58 @@ export function createProgram(rootNamesOrOptions: readonly string[] | CreateProg
         return result;
     }
 
-    function getCreateSourceFileOptions(fileName: string, moduleResolutionCache: ModuleResolutionCache | undefined, host: CompilerHost, options: CompilerOptions): CreateSourceFileOptions {
+    function getImpliedNodeFormatForFile(fileName: string) {
+        // Map to an output file before looking up a package.json that might
+        // contain `"type": "module"`. This allows the same input `.ts` file to be
+        // compiled as both CommonJS and ESM depending on the tsconfig's `outDir`.
+        let fileNameForModuleFormatDetection = getNormalizedAbsolutePath(fileName, currentDirectory);
+        if (moduleFormatNeedsPackageJsonLookup(fileName, options)) {
+            const projectReference = getResolvedProjectReferenceToRedirect(fileName);
+            if (projectReference) {
+                // We have a source file that is an input to a referenced project.
+                // This should only happen with `useSourceOfProjectReferenceRedirect`
+                // enabled (i.e. TS Server scenarios).
+                Debug.assert(useSourceOfProjectReferenceRedirect);
+                fileNameForModuleFormatDetection =
+                    getOutputDeclarationFileName(
+                        fileNameForModuleFormatDetection,
+                        projectReference.commandLine,
+                        !host.useCaseSensitiveFileNames()) ||
+                    getOutputJSFileName(
+                        fileNameForModuleFormatDetection,
+                        projectReference.commandLine,
+                        !host.useCaseSensitiveFileNames()) ||
+                    fileNameForModuleFormatDetection;
+            }
+            else {
+                fileNameForModuleFormatDetection =
+                    getOutputDeclarationFileNameWithoutConfigFile(
+                        fileNameForModuleFormatDetection,
+                        options,
+                        !host.useCaseSensitiveFileNames(),
+                        currentDirectory,
+                        getAssumedCommonSourceDirectory) ||
+                    getOutputJSFileNameWithoutConfigFile(
+                        fileNameForModuleFormatDetection,
+                        options,
+                        !host.useCaseSensitiveFileNames(),
+                        currentDirectory,
+                        getAssumedCommonSourceDirectory) ||
+                    fileNameForModuleFormatDetection;

```

Comment:
I considered checking to see if a result computed from `declarationDir` disagrees with a result computed from `outDir` and issue an error, but that seems like such an edge case that it wouldn’t be worth spending time doing that for every source file that could be influenced by it. I decided to prefer `declarationDir` since that seems most likely to give a library author the same analysis that their users will see. But if someone tries really hard, they can make an invalid situation by using separate `outDir` and `declarationDir` with conflicting package.json files in each.